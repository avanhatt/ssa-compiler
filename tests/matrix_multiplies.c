#include <stdio.h>
#include <math.h>

void _p_multiply(double a[3][3], double b[3][3], double res1[3][3], double res2[3][3], double res3[3][3], double res4[3][3]) { 
    for (int i = 0; i < 3; i++) { 
        for (int j = 0; j < 3; j++) { 
            res1[i][j] = 0; 
            res2[i][j] = 0; 
            res3[i][j] = 0; 
            for (int k = 0; k < 3; k++) {
                res1[i][j] += sqrt(a[i][k]*b[k][j]);     
                res2[i][j] += sqrt(a[i][k])*b[k][j];      
                res3[i][j] += a[i][k]*sqrt(b[k][j]);     
            }
        } 
    } 
    for (int i = 0; i < 3; i++) { 
        for (int j = 0; j < 3; j++) { 
            res4[i][j] = res1[i][j] + res2[i][j] + res3[i][j];
        } 
    } 
} 

int main(int argc, char const *argv[]) {
    double res1[3][3];
    double res2[3][3];
    double res3[3][3];
    double res4[3][3];
    double a[3][3] = {{1, 1, 1}, {2, 2, 2}, {3, 3, 3}}; 
    double b[3][3] = {{1, 2, 3}, {1, 2, 3}, {1, 2, 3}}; 
  
    _p_multiply(a, b, res1, res2, res3, res4); 

    for (int i = 0; i < 3; i++) { 
        for (int j = 0; j < 3; j++) {
            printf("%f ", res4[i][j]); 
        }
        printf("\n"); 
    } 

    return 0;
}