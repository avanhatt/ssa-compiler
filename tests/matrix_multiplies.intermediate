; ModuleID = '<stdin>'
source_filename = "tests/matrix_multiplies.c"
target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-macosx10.14.0"

@__const.main.a = private unnamed_addr constant [3 x [3 x double]] [[3 x double] [double 1.000000e+00, double 1.000000e+00, double 1.000000e+00], [3 x double] [double 2.000000e+00, double 2.000000e+00, double 2.000000e+00], [3 x double] [double 3.000000e+00, double 3.000000e+00, double 3.000000e+00]], align 16
@__const.main.b = private unnamed_addr constant [3 x [3 x double]] [[3 x double] [double 1.000000e+00, double 2.000000e+00, double 3.000000e+00], [3 x double] [double 1.000000e+00, double 2.000000e+00, double 3.000000e+00], [3 x double] [double 1.000000e+00, double 2.000000e+00, double 3.000000e+00]], align 16
@.str = private unnamed_addr constant [4 x i8] c"%f \00", align 1

; Function Attrs: nounwind ssp uwtable
define void @multiply([3 x double]* nocapture readonly, [3 x double]* nocapture readonly, [3 x double]* nocapture, [3 x double]* nocapture, [3 x double]* nocapture, [3 x double]* nocapture) local_unnamed_addr #0 {
  br label %7

; <label>:7:                                      ; preds = %41, %6
  %8 = phi i64 [ 0, %6 ], [ %42, %41 ], !partition !3, !start !3, !end !3
  br label %9

; <label>:9:                                      ; preds = %38, %7
  %10 = phi i64 [ 0, %7 ], [ %39, %38 ], !partition !4, !start !5, !end !5
  %11 = getelementptr inbounds [3 x double], [3 x double]* %2, i64 %8, i64 %10, !partition !3, !start !4, !end !6
  store double 0.000000e+00, double* %11, align 8, !tbaa !7, !partition !3, !start !6, !end !5
  %12 = getelementptr inbounds [3 x double], [3 x double]* %3, i64 %8, i64 %10, !partition !3, !start !5, !end !11
  store double 0.000000e+00, double* %12, align 8, !tbaa !7, !partition !3, !start !11, !end !12
  %13 = getelementptr inbounds [3 x double], [3 x double]* %4, i64 %8, i64 %10, !partition !4, !start !4, !end !6
  store double 0.000000e+00, double* %13, align 8, !tbaa !7, !partition !4, !start !6, !end !5
  br label %14

; <label>:14:                                     ; preds = %14, %9
  %15 = phi i64 [ 0, %9 ], [ %36, %14 ], !partition !3, !start !13, !end !13
  %16 = getelementptr inbounds [3 x double], [3 x double]* %0, i64 %8, i64 %15, !partition !4, !start !14, !end !12
  %17 = load double, double* %16, align 8, !tbaa !7, !partition !4, !start !15, !end !16
  %18 = getelementptr inbounds [3 x double], [3 x double]* %1, i64 %15, i64 %10, !partition !4, !start !3, !end !13
  %19 = load double, double* %18, align 8, !tbaa !7, !partition !4, !start !17, !end !18
  %20 = fmul double %17, %19, !partition !3, !start !19, !end !20
  %21 = tail call double @llvm.sqrt.f64(double %20), !partition !4, !start !18, !end !21
  %22 = load double, double* %11, align 8, !tbaa !7, !partition !3, !start !22, !end !11
  %23 = fadd double %22, %21, !partition !3, !start !16, !end !23
  store double %23, double* %11, align 8, !tbaa !7, !partition !3, !start !14, !end !22
  %24 = load double, double* %16, align 8, !tbaa !7, !partition !4, !start !12, !end !15
  %25 = tail call double @llvm.sqrt.f64(double %24), !partition !3, !start !3, !end !13
  %26 = load double, double* %18, align 8, !tbaa !7, !partition !4, !start !24, !end !23
  %27 = fmul double %25, %26, !partition !3, !start !6, !end !14
  %28 = load double, double* %12, align 8, !tbaa !7, !partition !3, !start !11, !end !12
  %29 = fadd double %28, %27, !partition !3, !start !12, !end !16
  store double %29, double* %12, align 8, !tbaa !7, !partition !3, !start !20, !end !25
  %30 = load double, double* %16, align 8, !tbaa !7, !partition !4, !start !23, !end !26
  %31 = load double, double* %18, align 8, !tbaa !7, !partition !4, !start !26, !end !17
  %32 = tail call double @llvm.sqrt.f64(double %31), !partition !3, !start !23, !end !18
  %33 = fmul double %30, %32, !partition !3, !start !18, !end !27
  %34 = load double, double* %13, align 8, !tbaa !7, !partition !4, !start !16, !end !24
  %35 = fadd double %34, %33, !partition !3, !start !27, !end !19
  store double %35, double* %13, align 8, !tbaa !7, !partition !4, !start !20, !end !25
  %36 = add nuw nsw i64 %15, 1, !partition !4, !start !21, !end !20
  %37 = icmp eq i64 %36, 3, !partition !4, !start !13, !end !14
  br i1 %37, label %38, label %14

; <label>:38:                                     ; preds = %14
  %39 = add nuw nsw i64 %10, 1, !partition !3, !start !3, !end !13
  %40 = icmp eq i64 %39, 3, !partition !4, !start !3, !end !13
  br i1 %40, label %41, label %9

; <label>:41:                                     ; preds = %38
  %42 = add nuw nsw i64 %8, 1, !partition !3, !start !4, !end !6
  %43 = icmp eq i64 %42, 3, !partition !4, !start !3, !end !13
  br i1 %43, label %44, label %7

; <label>:44:                                     ; preds = %41, %59
  %45 = phi i64 [ %60, %59 ], [ 0, %41 ], !partition !3, !start !3, !end !3
  br label %46

; <label>:46:                                     ; preds = %46, %44
  %47 = phi i64 [ 0, %44 ], [ %57, %46 ], !partition !4, !start !26, !end !26
  %48 = getelementptr inbounds [3 x double], [3 x double]* %2, i64 %45, i64 %47, !partition !3, !start !4, !end !6
  %49 = load double, double* %48, align 8, !tbaa !7, !partition !3, !start !6, !end !5
  %50 = getelementptr inbounds [3 x double], [3 x double]* %3, i64 %45, i64 %47, !partition !4, !start !13, !end !14
  %51 = load double, double* %50, align 8, !tbaa !7, !partition !4, !start !23, !end !26
  %52 = fadd double %49, %51, !partition !3, !start !22, !end !12
  %53 = getelementptr inbounds [3 x double], [3 x double]* %4, i64 %45, i64 %47, !partition !4, !start !12, !end !24
  %54 = load double, double* %53, align 8, !tbaa !7, !partition !4, !start !24, !end !23
  %55 = fadd double %52, %54, !partition !3, !start !5, !end !22
  %56 = getelementptr inbounds [3 x double], [3 x double]* %5, i64 %45, i64 %47, !partition !3, !start !12, !end !24
  store double %55, double* %56, align 8, !tbaa !7, !partition !3, !start !24, !end !23
  %57 = add nuw nsw i64 %47, 1, !partition !4, !start !14, !end !12
  %58 = icmp eq i64 %57, 3, !partition !4, !start !3, !end !13
  br i1 %58, label %59, label %46

; <label>:59:                                     ; preds = %46
  %60 = add nuw nsw i64 %45, 1, !partition !3, !start !4, !end !6
  %61 = icmp eq i64 %60, 3, !partition !4, !start !3, !end !13
  br i1 %61, label %62, label %44

; <label>:62:                                     ; preds = %59
  ret void, !partition !3, !start !3, !end !4
}

; Function Attrs: nounwind readnone speculatable
declare double @llvm.sqrt.f64(double) #1

; Function Attrs: nounwind ssp uwtable
define i32 @main(i32, i8** nocapture readnone) local_unnamed_addr #0 {
  %3 = alloca [3 x [3 x double]], align 16
  %4 = alloca [3 x [3 x double]], align 16
  %5 = alloca [3 x [3 x double]], align 16
  %6 = alloca [3 x [3 x double]], align 16
  %7 = alloca [3 x [3 x double]], align 16
  %8 = alloca [3 x [3 x double]], align 16
  %9 = bitcast [3 x [3 x double]]* %7 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull align 16 %9, i8* align 16 bitcast ([3 x [3 x double]]* @__const.main.a to i8*), i64 72, i1 false)
  %10 = bitcast [3 x [3 x double]]* %8 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* nonnull align 16 %10, i8* align 16 bitcast ([3 x [3 x double]]* @__const.main.b to i8*), i64 72, i1 false)
  %11 = getelementptr inbounds [3 x [3 x double]], [3 x [3 x double]]* %7, i64 0, i64 0
  %12 = getelementptr inbounds [3 x [3 x double]], [3 x [3 x double]]* %8, i64 0, i64 0
  %13 = getelementptr inbounds [3 x [3 x double]], [3 x [3 x double]]* %3, i64 0, i64 0
  %14 = getelementptr inbounds [3 x [3 x double]], [3 x [3 x double]]* %4, i64 0, i64 0
  %15 = getelementptr inbounds [3 x [3 x double]], [3 x [3 x double]]* %5, i64 0, i64 0
  %16 = getelementptr inbounds [3 x [3 x double]], [3 x [3 x double]]* %6, i64 0, i64 0
  call void @multiply([3 x double]* nonnull %11, [3 x double]* nonnull %12, [3 x double]* nonnull %13, [3 x double]* nonnull %14, [3 x double]* nonnull %15, [3 x double]* nonnull %16)
  br label %17

; <label>:17:                                     ; preds = %26, %2
  %18 = phi i64 [ 0, %2 ], [ %28, %26 ]
  br label %19

; <label>:19:                                     ; preds = %19, %17
  %20 = phi i64 [ 0, %17 ], [ %24, %19 ]
  %21 = getelementptr inbounds [3 x [3 x double]], [3 x [3 x double]]* %6, i64 0, i64 %18, i64 %20
  %22 = load double, double* %21, align 8, !tbaa !7
  %23 = tail call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @.str, i64 0, i64 0), double %22)
  %24 = add nuw nsw i64 %20, 1
  %25 = icmp eq i64 %24, 3
  br i1 %25, label %26, label %19

; <label>:26:                                     ; preds = %19
  %27 = tail call i32 @putchar(i32 10)
  %28 = add nuw nsw i64 %18, 1
  %29 = icmp eq i64 %28, 3
  br i1 %29, label %30, label %17

; <label>:30:                                     ; preds = %26
  ret i32 0
}

; Function Attrs: argmemonly nounwind
declare void @llvm.memcpy.p0i8.p0i8.i64(i8* nocapture writeonly, i8* nocapture readonly, i64, i1) #2

; Function Attrs: nounwind
declare i32 @printf(i8* nocapture readonly, ...) local_unnamed_addr #3

; Function Attrs: nounwind
declare i32 @putchar(i32) local_unnamed_addr #4

attributes #0 = { nounwind ssp uwtable "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "less-precise-fpmad"="false" "min-legal-vector-width"="0" "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="false" "stack-protector-buffer-size"="8" "target-cpu"="penryn" "target-features"="+cx16,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+sse4.1,+ssse3,+x87" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #1 = { nounwind readnone speculatable }
attributes #2 = { argmemonly nounwind }
attributes #3 = { nounwind "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "less-precise-fpmad"="false" "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="false" "stack-protector-buffer-size"="8" "target-cpu"="penryn" "target-features"="+cx16,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+sse4.1,+ssse3,+x87" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #4 = { nounwind }

!llvm.module.flags = !{!0, !1}
!llvm.ident = !{!2}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"PIC Level", i32 2}
!2 = !{!"clang version 8.0.1 (tags/RELEASE_801/final)"}
!3 = !{!"0"}
!4 = !{!"1"}
!5 = !{!"5"}
!6 = !{!"4"}
!7 = !{!8, !8, i64 0}
!8 = !{!"double", !9, i64 0}
!9 = !{!"omnipotent char", !10, i64 0}
!10 = !{!"Simple C/C++ TBAA"}
!11 = !{!"8"}
!12 = !{!"9"}
!13 = !{!"3"}
!14 = !{!"6"}
!15 = !{!"10"}
!16 = !{!"11"}
!17 = !{!"15"}
!18 = !{!"16"}
!19 = !{!"20"}
!20 = !{!"22"}
!21 = !{!"19"}
!22 = !{!"7"}
!23 = !{!"13"}
!24 = !{!"12"}
!25 = !{!"23"}
!26 = !{!"14"}
!27 = !{!"18"}
