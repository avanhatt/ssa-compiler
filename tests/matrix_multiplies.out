9.000000 14.485281 19.392305 
14.485281 22.970563 30.468696 
19.392305 30.468696 40.176915 
